### PR TITLE
feature-74: Go to the video excerpt corresponding to a comment

### PR DESCRIPTION
**Pull request by astrid-devericourt**
We, Antoine-Valentin Charpentier, Marina Peralta-Fonseca, Astrid Royer de Véricourt, Sun Sun, Agathe Allirol--Binet, hereby grant to Hyperglosae maintainers the right to publish our contribution under the terms of any licenses the Free Software Foundation classifies as Free Software Licenses.

### CODE REVIEW

File: features/focus_on_glose.feature [hunk 1]
```diff
@@ -33,4 +33,3 @@ Scénario: dans le cas d'une comparaison entre œuvres d'art apparentées
   Et "Photographie : vitrail, baie 6, Église Saint-Martin-ès-Vignes, Troyes" une des sources
   Et l'image intégrée en source "SNZ 113" est légendée par son titre : "Photographie : vitrail, baie 113, Église Saint-Nizier, Troyes"
   Et l'image intégrée en source "SMV 006 Lavage" est légendée par son titre : "Photographie : vitrail, baie 6, Église Saint-Martin-ès-Vignes, Troyes"
-
```
Comment (benel) [inline features/focus_on_glose.feature:36]: Please don't insert lines that are not related with your feature.

File: features/step_definitions/outcome.rb [hunk 2]
```diff
@@ -71,4 +71,9 @@
 
 Alors("{string} est le document affiché à l'écran") do |text|
   expect(find('.runningHead')).to have_content text
+end
+
+Alors('la vidéo du document principal {string} se lance de {string} secondes à {string} secondes') do |url, start, ending|
+  path = url + '?start='+ start +'&end=' + ending + '&autoplay=1&mute=1'
+  expect(find('iframe')['src']).to include(path)
```
Comment (benel) [inline features/step_definitions/outcome.rb:78]: This step doesn't pass on GitHub actions. Did it pass on your laptop?

If I remember right, members of your team (@AVCharpentier and @marina-pera) already faced the same problem and found a solution by using `within` (see https://github.com/Hypertopic/HyperGlosae/blob/main/features/step_definitions/outcome.rb#L62-L66).

File: frontend/src/Comment.js [hunk 3]
```diff
@@ -0,0 +1,35 @@
+import './comment.css';
+
+function Comment({ children }) {
+  const regExp = /(\d{2}:\d{2}:\d{2}\.\d{3}\s*-->\s*\d{2}:\d{2}:\d{2}\.\d{3})/;
```
Comment (benel) [inline frontend/src/Comment.js:4]: You may change the constant name in order to tell that this is a time code format.

File: frontend/src/FormattedText.js [hunk 4]
```diff
@@ -2,14 +2,16 @@ import ReactMarkdown from 'react-markdown';
...
 
 function FormattedText({children}) {
   return (
     <ReactMarkdown
       remarkPlugins={[remarkDefinitionList, remarkUnwrapImages]}
       components={{
         img: (x) => embedVideo(x) || CroppedImage(x),
-        a: ({children, href}) => <a href={href}>{children}</a>
+        a: ({children, href}) => <a href={href}>{children}</a>,
+        p: (x) => Comment(x) || <p>{x.children}</p>
```
Comment (benel) [inline frontend/src/FormattedText.js:14]: You may insert `p` before `a`, so that you don't get the `blame` of the line with `a`.

File: frontend/src/Comment.js [hunk 5]
```diff
@@ -0,0 +1,35 @@
+import './comment.css';
+
+function Comment({ children }) {
```
Comment (benel) [inline frontend/src/Comment.js:3]: The name of your component is very general. Are you sure it is handling "comments" of any types?

File: frontend/src/Comment.js [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+    let startTime = Number(hour * 3600) + Number(min * 60) + Number(sec);
+    [hour, min, sec] = end.split(/[:.]/);
+    let endTime = Number(hour * 3600) + Number(min * 60) + Number(sec);
+    let iframe = document.getElementsByTagName('iframe');
+    if (iframe.length != 0) {
+      let youTubeLink = new URL(iframe[0].src);
+      let youTubeBaseLink = youTubeLink.origin + youTubeLink.pathname;
+      let targetLink = `${youTubeBaseLink}?start=${startTime}&end=${endTime}&autoplay=1&mute=1`;
+      iframe[0].src = targetLink;
+      console.log(targetLink);
```
Comment (benel) [inline frontend/src/Comment.js:30]: Please don't commit debug logs.

File: frontend/src/Comment.js [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+    return <p
+      onClick={
+        e => {
+          e.preventDefault();
+          e.stopPropagation();
+          playVideoAt(children[0]);
+        }
+      }
+      className="comment"
+      title="Click to jump"
```
Comment (benel) [inline frontend/src/Comment.js:15]: Watching movies have never been so athletic ;)
Please find a better description.

**Review by benel — CHANGES_REQUESTED** [review]
Thank you very for this important contribution @Hypertopic/histoire-du-cinema.

Please edit the following lines so that I can merge your pull request.

File: features/step_definitions/outcome.rb (hunk 2 above)
Comment (AVCharpentier) [reply features/step_definitions/outcome.rb]: The test works properly on our computers without using within_frame.
image

With within_frame :
image
This is normal that this second version does not work on our computers and on GitHub. Here we change the context of research. We search in the content of the iframe, that is to say on the youtube page and not on the Hyperglossae site.

Note that the youtube page does not include the beginning and end of the extract. The management of video playback is processed by a javascript script and not stored in an HTML element, so complicated to make a classic find to perform a search in an HTML element. 
image

Then I modified the test to search JS scripts.
image

I am not sure that relying on the youtube page is the best solution because Youtube changes their structure very regularly which could lead to the test not working later.

I hope this modification will fix github issue

File: frontend/src/Comment.js (hunk 7 above)
Comment (AVCharpentier) [reply frontend/src/Comment.js]: Done

File: frontend/src/Comment.js (hunk 6 above)
Comment (AVCharpentier) [reply frontend/src/Comment.js]: Done

File: frontend/src/Comment.js (hunk 5 above)
Comment (AVCharpentier) [reply frontend/src/Comment.js]: Done

File: frontend/src/FormattedText.js (hunk 4 above)
Comment (AVCharpentier) [reply frontend/src/FormattedText.js]: Done

File: frontend/src/Comment.js (hunk 3 above)
Comment (AVCharpentier) [reply frontend/src/Comment.js]: Done

File: features/step_definitions/outcome.rb (hunk 2 above)
Comment (AVCharpentier) [reply features/step_definitions/outcome.rb]: Obviously the tests are still failing on GitHub. I wonder if the problem is more of a commit order nature. Maybe it runs the tests commit by commit without worrying about next commits.

I wonder this because the tests are essentially based on the commit of samples. If the samples are not good, tests based on them will not work.

The failed test error would come from there?

**Comment by AVCharpentier** [comment]
I might have an idea of why it works on our computers but not on GitHub. It's probably because we haven't committed the changes to the sample that includes the video; at least, I don't see it. On Hyperglosae, the video is still in the form of a clickable image. I haven't tried making the modifications yet to see if it works with a real video.

**Comment by benel** [comment]
Merged as e4cfcf2bcb9ecfcda50a73e044019778830a6218.

Thank you for your contribution @cug-sun, @marina-pera, @astrid-devericourt, @GagatheRollex, @AVCharpentier !
